File: android/src/main/kotlin/cn/mryt/sxws/mryt_share_plugin/MrytSharePlugin.kt
package cn.mryt.sxws.mryt_share_plugin

import android.annotation.SuppressLint
import android.app.Activity
import cn.mryt.sxws.mryt_share_plugin.enumerate.ShareMethodEnum
import cn.mryt.sxws.mryt_share_plugin.manager.PermissionManager
import cn.mryt.sxws.mryt_share_plugin.manager.ShareSdkManager
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry.Registrar

@SuppressLint("StaticFieldLeak")

class MrytSharePlugin : MethodCallHandler {


    companion object {

        var mActivity: Activity? = null
        var mChannel: MethodChannel? = null
        @JvmStatic
        fun registerWith(registrar: Registrar) {
            mChannel = MethodChannel(registrar.messenger(), "mryt_share_plugin")
            mChannel?.setMethodCallHandler(MrytSharePlugin())
            mActivity = registrar.activity()
            // onActivityResult
            ShareSdkManager.instance.setActivityResultListener(mActivity,registrar)

            // onRequestPermissionsResult
            PermissionManager.instance.setRequestPermissionsResultListener(mActivity,registrar)

        }
    }

    override fun onMethodCall(call: MethodCall, result: Result) {

        when (call.method) {
            ShareMethodEnum.METHOD_INIT_SHARESDK -> {

                val appId: String? = call.argument("appId")
                ShareSdkManager.instance.initShareSdk(appId, result)

            }
            ShareMethodEnum.METHOD_SET_SHARE_PLATFORM -> {

                val platform: String? = call.argument("platform")
                val appKey: String? = call.argument("appKey")
                val appSecret: String? = call.argument("appSecret")
                val callBackUrl: String? = call.argument("callBackUrl")
                ShareSdkManager.instance.setSharePlatform(platform, appKey, appSecret, callBackUrl, result)
            }
            ShareMethodEnum.METHOD_SET_MINPROGRAMID -> {

                val platform: String? = call.argument("platform")
                val minProgramId: String? = call.argument("minProgramId")
                ShareSdkManager.instance.setMinProgramId(platform, minProgramId, result)

            }
            ShareMethodEnum.METHOD_SHARE_TEXT -> {

                val platform: String? = call.argument("platform")
                val text: String? = call.argument("text")
                ShareSdkManager.instance.shareText(mActivity, platform, text, result, mChannel)

            }
            ShareMethodEnum.METHOD_SHARE_ONEIMAGE -> {

                val platform: String? = call.argument("platform")
                val imgUrl: String? = call.argument("imgUrl")
                val description: String? = call.argument("description")
                ShareSdkManager.instance.shareOneImage(mActivity, platform, imgUrl, description, result, mChannel)

            }
            ShareMethodEnum.METHOD_SHARE_IMAGELIST -> {

                val platform: String? = call.argument("platform")
                val imgListUrl: ArrayList<String>? = call.argument("imgListUrl")
                val description: String? = call.argument("description")
                ShareSdkManager.instance.shareImageList(mActivity, platform, imgListUrl, description, result, mChannel)

            }
            ShareMethodEnum.METHOD_SHARE_WEB -> {

                val platform: String? = call.argument("platform")
                val webUrl: String? = call.argument("webUrl")
                val webTitle: String? = call.argument("webTitle")
                val webDes: String? = call.argument("webDes")
                val webThumb: String? = call.argument("webThumb")
                ShareSdkManager.instance.shareWeb(mActivity, platform, webUrl, webTitle, webDes, webThumb, result, mChannel)

            }
            ShareMethodEnum.METHOD_SHARE_VIDEO -> {

                val platform: String? = call.argument("platform")
                val videoUrl: String? = call.argument("videoUrl")
                val videoTitle: String? = call.argument("videoTitle")
                val videoDes: String? = call.argument("videoDes")
                val videoThumb: String? = call.argument("videoThumb")
                ShareSdkManager.instance.shareVideo(mActivity, platform, videoUrl, videoTitle, videoDes, videoThumb, result, mChannel)

            }
            ShareMethodEnum.METHOD_SHARE_MINPROGRAM -> {

                val platform: String? = call.argument("platform")
                val miniWebUrl: String? = call.argument("miniWebUrl")
                val miniPagePath: String? = call.argument("miniPagePath")
                val miniTitle: String? = call.argument("miniTitle")
                val miniDes: String? = call.argument("miniDes")
                val miniThumb: String? = call.argument("miniThumb")
                ShareSdkManager.instance.shareMinProgram(mActivity, platform, miniWebUrl, miniPagePath, miniTitle, miniDes, miniThumb, result, mChannel)

            }
            else -> {
                result.notImplemented()
            }
        }

    }
}

File: ios/Classes/MrytSharePlugin.m
//
//  MrytSharePlugin.m
//  Runner
//
//  Created by 武贤业 on 2019/10/7.
//  Copyright © 2019 The Chromium Authors. All rights reserved.
//

#import "MrytSharePlugin.h"
#import <UMShare/UMShare.h>
#import <UMCommon/UMCommon.h>
#import <UMCommon/UMConfigure.h>

static NSString *initShareSdk = @"initShareSdk";
static NSString *setSharePlatform = @"setSharePlatform";
static NSString *shareText = @"shareText";
static NSString *shareImageList = @"shareImageList";
static NSString *shareWeb = @"shareWeb";
static NSString *shareVideo = @"shareVideo";
static NSString *shareMinProgram = @"shareMinProgram";
static NSString *shareOneImage = @"shareOneImage";
static NSString *onShareStart = @"onShareStart";
static NSString *onShareSuccess = @"onShareSuccess";
static NSString *onShareError = @"onShareError";
static NSString *onShareCancel = @"onShareCancel";


@interface NSError (MrytSharePlugin)

@property (nonatomic, readonly) FlutterError *flutterError;

@end

@implementation NSError (MrytSharePlugin)

- (FlutterError *)flutterError {
    return [FlutterError errorWithCode:[NSString
                                        stringWithFormat:@"Error %d", (int)self.code]
                               message:self.domain
                               details:self.localizedRecoverySuggestion];
}

@end


@implementation MrytSharePlugin

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
    FlutterMethodChannel *channel = [FlutterMethodChannel methodChannelWithName:@"mryt_share_plugin" binaryMessenger:[registrar messenger]];
    MrytSharePlugin *plugin = [[MrytSharePlugin alloc] init];
    plugin.channel = channel;
    
    [registrar addApplicationDelegate:plugin];
    [registrar addMethodCallDelegate:plugin channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall *)call
                  result:(FlutterResult)result {
    NSString *method = call.method;
    if ([method isEqualToString:initShareSdk]) {
        [self initShareSdk:call result:result];
    } else if ([method isEqualToString:setSharePlatform]) {
        [self setSharePlatform:call result:result];
    } else if ([method isEqualToString:shareText]) {
        [self shareText:call result:result];
    } else if ([method isEqualToString:shareImageList]) {
        [self shareImageList:call result:result];
    } else if ([method isEqualToString:shareWeb]) {
        [self shareWeb:call result:result];
    } else if ([method isEqualToString:shareVideo]) {
        [self shareVideo:call result:result];
    } else if ([method isEqualToString:shareMinProgram]) {
        [self shareMinProgram:call result:result];
    } else if ([method isEqualToString:shareOneImage]) {
        [self shareOneImage:call result:result];
    } else {
        result(FlutterMethodNotImplemented);
    }
}

#pragma mark - dart调用

- (void)initShareSdk:(FlutterMethodCall *)call
              result:(FlutterResult)result  {
    NSDictionary *dic = call.arguments;
    NSString *appId = dic[@"appId"];
    if (appId == nil) {
        result([self normalFlutterError]);
        return;
    }
    
    NSString *channel = dic[@"channel"];
    [UMConfigure initWithAppkey:appId
                        channel:channel ?: @""];
}

- (void)setSharePlatform:(FlutterMethodCall *)call
                  result:(FlutterResult)result {
    NSDictionary *dic = call.arguments;
    NSString *appKey = dic[@"appKey"];
    NSString *platform = dic[@"platform"];
    NSString *appSecret = dic[@"appSecret"];

    if (appKey == nil ||
        platform == nil ||
        appSecret == nil) {
        result([self normalFlutterError]);
        return;
    }
    
    NSString *callBackUrl = dic[@"callBackUrl"];
    
    BOOL success = [[UMSocialManager defaultManager] setPlaform:[self getSharePlatformType:platform]
                                                         appKey:appKey
                                                      appSecret:appSecret
                                                    redirectURL:callBackUrl];
    result(@{@"success": @(success)});
}

- (void)shareText:(FlutterMethodCall *)call
           result:(FlutterResult)result {
    NSDictionary *dic = call.arguments;
    NSString *platform = dic[@"platform"];
    NSString *text = dic[@"text"];
    
    if (platform == nil || text == nil) {
        [self shareParameterError];
        return;
    }
    
    UMShareObject *shareObjc = [UMShareObject new];
    shareObjc.title = text;
    
    [self shareWithPlatform:platform
                shareObject:shareObjc
                     result:result];
}

- (void)shareOneImage:(FlutterMethodCall *)call
               result:(FlutterResult)result {
    NSDictionary *dic = call.arguments;
    NSString *platform = dic[@"platform"];
    NSString *imgUrl = dic[@"imgUrl"];
    
    if (platform == nil || imgUrl == nil) {
        [self shareParameterError];
        return;
    }
    
    NSString *desc = dic[@"desc"];
    
    UMShareImageObject *shareObjc = [UMShareImageObject new];
    shareObjc.shareImage = imgUrl;
    shareObjc.descr = desc;
    
    [self shareWithPlatform:platform
                shareObject:shareObjc
                     result:result];
}

- (void)shareImageList:(FlutterMethodCall *)call
                result:(FlutterResult)result {
    NSDictionary *dic = call.arguments;
    NSString *platform = dic[@"platform"];
    NSArray<FlutterStandardTypedData *> *imgDataList = dic[@"imgDataList"];
    NSMutableArray *arr = [NSMutableArray array];
    [imgDataList enumerateObjectsUsingBlock:^(FlutterStandardTypedData * _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        if (obj.data != nil) {
            [arr addObject:obj.data];
        }
    }];
    if (platform == nil || arr == nil || arr.count <= 0) {
        [self shareParameterError];
        return;
    }
    
    NSString *desc = dic[@"desc"];
    
    UMShareImageObject *shareObjc = [UMShareImageObject new];
    shareObjc.shareImageArray = [arr copy];
    shareObjc.descr = desc;

    
    [self shareWithPlatform:platform
                shareObject:shareObjc
                     result:result];

}

- (void)shareWeb:(FlutterMethodCall *)call
          result:(FlutterResult)result {
    NSDictionary *dic = call.arguments;
    NSString *platform = dic[@"platform"];
    NSString *webUrl = dic[@"webUrl"];
    
    if (platform == nil || webUrl == nil) {
        [self shareParameterError];
        return;
    }
    
    
    UMShareWebpageObject *shareObjc = [UMShareWebpageObject new];
    shareObjc.webpageUrl = webUrl;
    shareObjc.descr = dic[@"webDes"];
    shareObjc.title = dic[@"webTitle"];
    shareObjc.thumbImage = dic[@"webThumb"];

    
    [self shareWithPlatform:platform
                shareObject:shareObjc
                     result:result];

}

- (void)shareVideo:(FlutterMethodCall *)call
            result:(FlutterResult)result {
    NSDictionary *dic = call.arguments;
    NSString *platform = dic[@"platform"];
    NSString *videoUrl = dic[@"videoUrl"];
    
    if (platform == nil || videoUrl == nil) {
        [self shareParameterError];
        return;
    }
    
    
    UMShareVideoObject *shareObjc = [UMShareVideoObject new];
    shareObjc.videoUrl = videoUrl;
    shareObjc.descr = dic[@"videoDes"];
    shareObjc.title = dic[@"videoTitle"];
    shareObjc.thumbImage = dic[@"videoThumb"];

    
    [self shareWithPlatform:platform
                shareObject:shareObjc
                     result:result];
}

- (void)shareMinProgram:(FlutterMethodCall *)call
                 result:(FlutterResult)result {
    NSDictionary *dic = call.arguments;
    NSString *platform = dic[@"platform"];
    NSString *miniWebUrl = dic[@"miniWebUrl"];
    NSString *userName = dic[@"userName"];

    if (platform == nil ||
        userName == nil ||
        miniWebUrl) {
        [self shareParameterError];
        return;
    }
    
    NSNumber *miniProgramType = dic[@"miniProgramType"];
    NSNumber *withShareTicket = dic[@"withShareTicket"];
    
    UMShareMiniProgramObject *shareObjc = [UMShareMiniProgramObject new];
    shareObjc.miniProgramType = [miniProgramType integerValue];
    shareObjc.userName = userName;
    shareObjc.webpageUrl = miniWebUrl;
    
    shareObjc.descr = dic[@"miniDes"];
    shareObjc.title = dic[@"miniTitle"];
    shareObjc.thumbImage = dic[@"miniThumb"];
    
    shareObjc.withShareTicket = [withShareTicket boolValue];
    shareObjc.path = dic[@"miniPagePath"];

    
    [self shareWithPlatform:platform
                shareObject:shareObjc
                     result:result];

}


#pragma mark - FlutterApplicationLifeCycleDelegate


- (BOOL)application:(UIApplication *)application handleOpenURL:(NSURL *)url {
    [[UMSocialManager defaultManager] handleOpenURL:url];
    return YES;
}

- (BOOL)application:(UIApplication *)application
            openURL:(NSURL *)url
            options:(NSDictionary<UIApplicationOpenURLOptionsKey,id> *)options {
    [[UMSocialManager defaultManager] handleOpenURL:url options:options];
    return YES;
}

- (BOOL)application:(UIApplication *)application
            openURL:(NSURL *)url
  sourceApplication:(NSString *)sourceApplication
         annotation:(id)annotation {
    [[UMSocialManager defaultManager] handleOpenURL:url
                                  sourceApplication:sourceApplication
                                         annotation:annotation];
    return YES;
}

#pragma mark - private

- (void)shareWithPlatform:(NSString *)platformStr
              shareObject:(UMShareObject *)shareObject
                   result:(FlutterResult)result {
    UIViewController *vc = [self getCurrentViewController];
    if (vc == nil) {
        [self shareParameterError];
        return;
    }
    
    UMSocialMessageObject *messageObjc = [UMSocialMessageObject new];
    if ([shareObject isMemberOfClass:[UMShareObject class]]) {
        messageObjc.text = shareObject.title;
    } else {
        messageObjc.shareObject = shareObject;
    }
    NSString *startJsonStr = [self dictionaryToJson:@{@"method": onShareStart}];
    [self.channel invokeMethod:onShareStart arguments:startJsonStr];
    
    [[UMSocialManager defaultManager] shareToPlatform:[self getSharePlatformType:platformStr]
                                        messageObject:messageObjc
                                currentViewController:vc
                                           completion:^(id shareResponse, NSError *error) {
        if (error != nil) {
            if (error.code == UMSocialPlatformErrorType_Cancel) {
                NSString *jsonStr = [self dictionaryToJson:@{@"method": onShareCancel}];
                [self.channel invokeMethod:onShareCancel
                                 arguments:jsonStr];
            } else {
                NSString *jsonStr = [self dictionaryToJson:@{
                    @"success": @NO,
                    @"method": onShareError,
                    @"msg": [NSString stringWithFormat:@"Error %ld", error.code]
                }];

                [self.channel invokeMethod:onShareError
                                 arguments:jsonStr];
            }
            
        } else {
            NSString *jsonStr = [self dictionaryToJson:@{
                @"success": @YES,
                @"method": onShareSuccess
            }];
            [self.channel invokeMethod:onShareSuccess
                             arguments:jsonStr];
        }
    
    }];
}

- (void)shareParameterError {
    NSString *jsonStr = [self dictionaryToJson:@{
        @"success": @NO,
        @"method": onShareError,
        @"msg": @"缺少必要参数"
    }];
    
    [self.channel invokeMethod:onShareError arguments:jsonStr];
}

- (UMSocialPlatformType)getSharePlatformType:(NSString *)platformStr {
    UMSocialPlatformType type = UMSocialPlatformType_UnKnown;
    if ([platformStr isEqualToString:@"QQ"]) {
        type = UMSocialPlatformType_QQ;
    } else if ([platformStr isEqualToString:@"QZONE"]) {
        type = UMSocialPlatformType_Qzone;
    } else if ([platformStr isEqualToString:@"WEIXIN_FRIEND"]) {
        type = UMSocialPlatformType_WechatSession;
    } else if ([platformStr isEqualToString:@"WEIXIN_CIRCLE"]) {
        type = UMSocialPlatformType_WechatTimeLine;
    } else if ([platformStr isEqualToString:@"SINA"]) {
        type = UMSocialPlatformType_Sina;
    }
    return type;
}

- (UIViewController *)getCurrentViewController {
    return [UIApplication sharedApplication].keyWindow.rootViewController;
}



- (FlutterError *)normalFlutterError {
    return [FlutterError errorWithCode:@"Error -10000001"
                               message:@"MrytSharePlugin 普通缺少参数错误"
                               details:nil];
}

- (NSString *)dictionaryToJson:(NSDictionary *)dic {
    NSError *parseError = nil;
    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:dic
                                                       options:NSJSONWritingPrettyPrinted
                                                         error:&parseError];
    if (parseError == nil) {
        return [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
    } else {
        return @"";
    }
    
}


@end
